### PR TITLE
Prune the column values before doing each broadcast.

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by tpoterba** [comment]
This PR adds a step in the execution of several MatrixIR nodes to prune the column values to the schema used in the IR before broadcasting them.

File: src/main/scala/is/hail/expr/ir/PruneDeadFields.scala [hunk 1]
```diff
@@ -77,6 +77,27 @@ object PruneDeadFields {
...
+    val originalCols = mv.colValues
+    val memo = Memo.empty[BaseType]
+    val valueIRCopy = valueIR.deepCopy()
+    val colDep = memoizeValueIR(valueIRCopy, valueIR.typ, memo)
+      .m.mapValues(_._2)
+      .getOrElse("sa", if (isArray) TArray(TStruct()) else TStruct())
+    log.info(s"minimized col values:\n  From: ${ matrixType.colType }\n  To: ${ colDep }")
+    val newColsType = if (isArray) colDep.asInstanceOf[TArray] else TArray(colDep)
+    val newIndexedSeq = Interpret[IndexedSeq[Annotation]](
+      upcast(Ref("values", TArray(matrixType.colType)), newColsType), Env.empty[(Any, Type)]
```
Comment (jbloom22) [inline src/main/scala/is/hail/expr/ir/PruneDeadFields.scala]: hard to read, move Env to next line

File: src/main/scala/is/hail/expr/ir/PruneDeadFields.scala [hunk 2]
```diff
@@ -77,6 +77,27 @@ object PruneDeadFields {
     }
   }
 
+  def minimizeColValues(mv: MatrixValue, valueIR: IR, isArray: Boolean = false): (Type, BroadcastIndexedSeq, IR) = {
+    val matrixType = mv.typ
+    val originalCols = mv.colValues
```
Comment (jbloom22) [inline src/main/scala/is/hail/expr/ir/PruneDeadFields.scala]: prefer `oldColValues`

File: src/main/scala/is/hail/expr/ir/PruneDeadFields.scala [hunk 3]
```diff
@@ -77,6 +77,27 @@ object PruneDeadFields {
...
+    val memo = Memo.empty[BaseType]
+    val valueIRCopy = valueIR.deepCopy()
+    val colDep = memoizeValueIR(valueIRCopy, valueIR.typ, memo)
+      .m.mapValues(_._2)
+      .getOrElse("sa", if (isArray) TArray(TStruct()) else TStruct())
+    log.info(s"minimized col values:\n  From: ${ matrixType.colType }\n  To: ${ colDep }")
```
Comment (jbloom22) [inline src/main/scala/is/hail/expr/ir/PruneDeadFields.scala]: consider setting `val oldColType = matrixType.colType` above as you use it in three places.
do you really want to log this even when no reduction is done?

File: src/main/scala/is/hail/expr/ir/MatrixIR.scala [hunk 4]
```diff
@@ -610,12 +610,14 @@ case class MatrixAggregateRowsByKey(child: MatrixIR, expr: IR) extends MatrixIR
 
     val nCols = prev.nCols
 
+    val (minColType, minColValues, rewriteIR) = PruneDeadFields.minimizeColValues(prev, expr)
```
Comment (jbloom22) [inline src/main/scala/is/hail/expr/ir/MatrixIR.scala]: I think `pruneColValues` is a clearer and more consistent name (minimizing is overloaded).

**Comment by tpoterba** [comment]
good comments, addressed.

File: src/main/scala/is/hail/expr/ir/MatrixIR.scala [hunk 5]
```diff
@@ -610,12 +610,14 @@ case class MatrixAggregateRowsByKey(child: MatrixIR, expr: IR) extends MatrixIR
 
     val nCols = prev.nCols
 
+    val (minColType, minColValues, rewriteIR) = PruneDeadFields.pruneColValues(prev, expr)
```
Comment (jbloom22) [inline src/main/scala/is/hail/expr/ir/MatrixIR.scala]: do you tend to use the prefix `min` or `new` or `pruned` elsewhere?

Comment (tpoterba) [reply src/main/scala/is/hail/expr/ir/MatrixIR.scala]: I think it's `min` in a lot of places. It's a synonym with `pruned`

**Pull request merged**